PC_Sort and PC_IsSorted

mbredif opened the pull request:
(This is a rebased version of #102, taking into account @pramsey's PCDIMENSION_LIST comment)

This pull request proposes 2 new interrelated functions:
- `PC_Sort( pcpatch, text[] ) : pcpatch`
- `PC_IsSorted( pcpatch, text[], boolean ) : boolean`
#### `PC_Sort( pa pcpatch, attr text[] ) : pcpatch`

(pgpointcloud/pointcloud#54, LI3DS/pointcloud#10) 

Lexicographically sort a patch `pa` on the given dimensions `attr`.

What is natively implemented ?
- [x]  Uncompressed
- [x]  Other compressions fall back to uncompression
- [ ]  Dimensional Uncompressed
- [ ]  Dimensional SigBits
- [ ]  Dimensional Zlib
- [ ]  Dimensional RLE
- [ ]  GHT
- [ ]  LAZPERF

#### `PC_IsSorted( pa pcpatch, attr text[], strict boolean default false ) : boolean`

(LI3DS/pointcloud#9)

Test whether a patch `pa` is lexicographically sorted on the given dimensions `attr`.
The `strict` option requires strict sorting (no duplicates).

What is natively implemented ?
- [x]  Uncompressed
- [x]  Other compressions fall back to uncompression
- [x]  Dimensional Uncompressed (single dimension)
- [ ]  Dimensional SigBits (single dimension)
- [ ]  Dimensional Zlib (single dimension)
- [x]  Dimensional RLE (single dimension)
- [ ]  Dimensional Uncompressed
- [ ]  Dimensional SigBits
- [ ]  Dimensional Zlib
- [ ]  Dimensional RLE
- [ ]  GHT
- [ ]  LAZPERF

#### Contributors
- functionalities : @mbredif
- tests : @pblottiere

mbredif commented:
@strk, @pramsey, would you consider merging this PR, addressing #54 ?

strk commented on lib/CMakeLists.txt:
```diff
@@ -39,7 +40,7 @@ set_target_properties (libpc-static
     PREFIX "lib"
     CLEAN_DIRECT_OUTPUT 1
-    COMPILE_FLAGS -fPIC
+    COMPILE_FLAGS "-fPIC -std=gnu89"
```
What is this needed for ?

mbredif replied:
for `qsort_r`
Unfortunately my implementation needs a context pointer to store the sorting dimensions and the standard `qsort` is not reentrant and seems to ask for a global variable to store the context.
Alternatives qsort_r or qsort_s appear to be non-standard. (sort_r)[noporpoise/sort_r], seems to be a relevant polyfill, but I have not tried it yet.

strk commented on lib/cunit/Makefile:
```diff
@@ -14,7 +14,8 @@ OBJS =	\
 	cu_pc_patch_ght.o \
-        cu_pc_patch_lazperf.o
+        cu_pc_patch_lazperf.o \
+	cu_pc_sort.o
```
indent change here ?

mbredif replied:
Ok, I will fix that

mbredif commented:
That PR is now rebased, including fallback support for lazperf through uncompression.
The COMPILE_FLAGS requirement has been dropped using [noporpoise/sort_r](http://github.com/noporpoise/sort_r) which provides a `qsort_r` polyfill under a  Public Domain license. I added the sort_r.h header under lib/sort_r, is that the way to go ?

strk commented:
Seems good to me, please add a README under lib/sort_r pointing to the github repo and/or the stackexchange question so we know where to look in case a fix is needed (maybe also in the .h file itself)

mbredif commented:
Done !

strk commented:
Thanks !

elemoine commented on lib/cunit/cu_pc_sort.c:
```diff
@@ -0,0 +1,390 @@
+    CU_TEST_INFO_NULL
+};
+
+CU_SuiteInfo sort_suite = {"sort", init_suite, clean_suite, sort_tests};
```
This will cause a segfault with cunit-2.3-1. See https://github.com/pgpointcloud/pointcloud/pull/124. I'll create a PR for that.

I'll also fix the indent in this file.

mbredif replied:
thanks !
